moshi/src/main/java/com/squareup/moshi/AdapterMethodsFactory.kt

```diff
@@ -0,0 +1,362 @@
+    type: Type,
+    annotations: Set<Annotation>,
+    moshi: Moshi
+  ): JsonAdapter<*>? {
+    val toAdapter = get(toAdapters, type, annotations)
+    val fromAdapter = get(fromAdapters, type, annotations)
+    if (toAdapter == null && fromAdapter == null) return null
+
+    val delegate: JsonAdapter<Any>? = if (toAdapter == null || fromAdapter == null) {
+      try {
+        moshi.nextAdapter(this, type, annotations)
+      } catch (e: IllegalArgumentException) {
+        val missingAnnotation = if (toAdapter == null) "@ToJson" else "@FromJson"
+        throw IllegalArgumentException(
+          "No $missingAnnotation adapter for ${type.toStringWithAnnotations(annotations)}",
+          e
+        )
+      }
+    } else {
+      null
+    }
+
+    toAdapter?.bind(moshi, this)
+    fromAdapter?.bind(moshi, this)
+
+    return object : JsonAdapter<Any>() {
+      override fun toJson(writer: JsonWriter, value: Any?) {
+        when {
+          toAdapter == null -> {
+            requireNotNull(delegate).toJson(writer, value)
+          }
+          !toAdapter.nullable && value == null -> {
+            writer.nullValue()
+          }
+          else -> {
+            try {
+              toAdapter.toJson(moshi, writer, value)
+            } catch (e: InvocationTargetException) {
+              val cause = e.cause
+              if (cause is IOException) throw cause
+              throw JsonDataException("$cause at ${writer.path}", cause)
+            }
+          }
+        }
+      }
+
+      override fun fromJson(reader: JsonReader): Any? {
+        return when {
+          fromAdapter == null -> {
+            requireNotNull(delegate).fromJson(reader)
+          }
+          !fromAdapter.nullable && reader.peek() == JsonReader.Token.NULL -> {
+            reader.nextNull<Any>()
+            null
+          }
+          else -> {
+            try {
+              fromAdapter.fromJson(moshi, reader)
+            } catch (e: InvocationTargetException) {
+              val cause = e.cause
+              if (cause is IOException) throw cause
+              throw JsonDataException("$cause at ${reader.path}", cause)
+            }
+          }
+        }
+      }
+
+      override fun toString() = "JsonAdapter$annotations($type)"
+    }
+  }
+
+  companion object {
+    fun get(adapter: Any): AdapterMethodsFactory {
+      val toAdapters = mutableListOf<AdapterMethod>()
+      val fromAdapters = mutableListOf<AdapterMethod>()
+
+      var clazz = adapter.javaClass
+      while (clazz != Any::class.java) {
+        for (declaredMethod in clazz.declaredMethods) {
+          if (declaredMethod.isAnnotationPresent(ToJson::class.java)) {
+            val toAdapter = toAdapter(adapter, declaredMethod)
+            val conflicting = get(toAdapters, toAdapter.type, toAdapter.annotations)
+            require(conflicting == null) {
+              "Conflicting @ToJson methods:\n    ${requireNotNull(conflicting).method}\n    ${toAdapter.method}"
```

Comment:
```suggestion
              "Conflicting @ToJson methods:\n    ${conflicting.method}\n    ${toAdapter.method}"
```

Why is `requireNotNull` necessary here? Shouldn't the require handle this?

Reply:
https://github.com/square/moshi/pull/1467/commits/25aac2c6eab7b1854a02028a92071be426c3f99f